aten/src/ATen/native/cpu/MultinomialKernel.cpp

```diff
@@ -20,27 +21,32 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
   int64_t n_dist = self.dim() > 1 ? self.size(-2) : 1;
 
   /* cumulative probability distribution vector */
-  Tensor cum_dist = at::empty({n_categories}, self.options());
+  Tensor cum_dist = at::empty({n_categories}, self.options().dtype<accscalar_t>());
+  /* copy of original distribution to recalculate cdf */
+  Tensor self_copy = at::empty({n_categories}, self.options().dtype<scalar_t>());
```

Comment:
Two comments here:
- For the `with_replacement=True` case, we don't need `self_copy`. Let's not try to use too much extra memory when unnecessary.
- We don't really need a full copy of self, we just need a bitmap of whether or not an element has already been chosen. It is fine to assume that the user has at most 64 dimensions and use a std::bitmap<64> for this check. TensorIteratorReduce makes this assumption as well; there should be a global constant somewhere for this number (64)

Reply:
Did you mean std::bitset? A 64 element limit sounds strange to me - wouldn't that mean that only 64 classes are supported? That would be definitely too restrictive. A dynamically allocated bitmap could be used and the recalculation of the sum could be omitted by just summing up until the random threshhold-point hab been found ... but it would only be a small improvement.

Reply:
@andreaskoepf, I did mean std::bitset and you're completely right, I got `n_categories` and the number of dimensions mixed up. Sorry about that!

Reply:
Using a dynamically allocated bit array (`std::vector<bool>`) would save some memory. i.e., if we have n_categories = 1024 and we're working with float tensors, then instead of 4 * 1024 = 4kb we'd only need 1024 * 1/8 = 128 bytes.